[Naoki Shum] iP

NaokiShum opened the pull request:
### DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))
DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html). 
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😄 

And it is **FREE**!

Features:

- [x] Managing Tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

NaokiShum committed:
test

NaokiShum committed:
test2

NaokiShum committed:
test3

NaokiShum committed:
Level-1

NaokiShum committed:
level-2

NaokiShum committed:
Added Task class

NaokiShum committed:
Added deadline

NaokiShum committed:
Added event

NaokiShum committed:
Added task

NaokiShum committed:
Added Todo

NaokiShum committed:
modified

NaokiShum committed:
added test case

NaokiShum committed:
A-TextUiTesting

NaokiShum committed:
Added some errors for incorrect inputs

NaokiShum committed:
Modified for level 6

NaokiShum committed:
Modified testcase

NaokiShum committed:
Modified for level 6

NaokiShum committed:
Add new class Save

NaokiShum committed:
Modify to add getter method

NaokiShum committed:
Modify to add getter method

NaokiShum committed:
modify for level-7

NaokiShum committed:
Added task

NaokiShum committed:
Add duke.txt

NaokiShum committed:
Modify for level-8

NaokiShum committed:
Merge branch 'branch-Level-8'

NaokiShum committed:
modify for level-7,level-8

NaokiShum committed:
Added Command class

NaokiShum committed:
Added TaskList class

NaokiShum committed:
Added Ui class

NaokiShum committed:
Added Storage class

NaokiShum committed:
Modified for more OOP

NaokiShum committed:
Added command class

NaokiShum committed:
Added tasklist class

NaokiShum committed:
Modified for more OOP

NaokiShum committed:
Added Ui class

NaokiShum committed:
Still in progress

NaokiShum committed:
Modified for more increments

NaokiShum committed:
Modified for packages

NaokiShum committed:
Added some J-Unit test

NaokiShum committed:
Package as JAR file

NaokiShum committed:
Added javadocs

NaokiShum committed:
Add javadoc

NaokiShum committed:
Update coding standard

NaokiShum committed:
Improve coding standard

NaokiShum committed:
modify for level-9

NaokiShum committed:
Merge tag 'A-JavaDoc'

NaokiShum committed:
Fix conflicts after merging

NaokiShum committed:
Fix indentation

NaokiShum committed:
Add javadocs

NaokiShum committed:
Add javsdocs

NaokiShum committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

NaokiShum committed:
Change mainClassName

zihaooo9 commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,132 @@
+ */
+public class Command {
+    /**
+     * executes different commands from user and save tasks into a file
```
The javadoc comment should have an empty line between the description and parameter section. Punctuation should also be behind each parameter description as well.

Example:
```java
/**
 * Returns lateral location of the specified position.
 * If the position is unset, NaN is returned.
 *
 * @param x  X coordinate of position.
 * @param y Y coordinate of position.
 * @param zone Zone of position.
 * @return Lateral location.
 * @throws IllegalArgumentException  If zone is <= 0.
 */
public double computeLocation(double x, double y, int zone)
    throws IllegalArgumentException {
  ...
}
```

I noticed the same issue in several other places too.

chownces replied:
Agreed! Also, I feel that the first word should be capitalized, as it would look more professional and proper (similarly for the rest of the JavaDoc comments in the repo)

zihaooo9 commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,132 @@
+
+        while (!hasExit) {
+            String command = Parser.parseCommand(sc.next());
+            switch(command){
```
There should be a white space after the `switch` statement. Generally, Java reserved words should be followed by a white space.

```suggestion
            switch(command) {
```

I noticed the same issue in several other places too.

zihaooo9 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+     * @param description takes in a String describing the task description
+     * @param by takes in a String representing the deadline of the task
+     */
+    public Deadline(String description,String by) {
```
Commas should be followed by a white space.

zihaooo9 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+     * @return a String representing the deadline
+     */
+    public String getBy(){
+        return this.by;
```
You can avoid unnecessary use of `this` with fields. Use the `this` keyword only when a field is shadowed by a method or constructor parameter. This helps to reduce unnecessary noise in your code!

zihaooo9 commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,11 @@
+package duke;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
I would suggest organising the import statements (see example below) so that it would be much more consistent when you add more import statements in the future!

Example: 
```
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import java.io.File;
import java.io.IOException;

import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBException;

import org.loadui.testfx.GuiTest;
import org.testfx.api.FxToolkit;

import com.google.common.io.Files;

import javafx.geometry.Bounds;
import javafx.geometry.Point2D;
import junit.framework.AssertionFailedError;
```

zihaooo9 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,131 @@
+     * @return a String representing text to be added
+     */
+    private static String textToAdd(Task t) {
+        String mes = "";
```
Perhaps a more intuitive variable name here?

zihaooo9 reviewed (commented):
Code generally looks good to me apart from the issues that i have pointed out! All the best for the remaining development of Duke!

NaokiShum committed:
Add GUI

NaokiShum committed:
Merge branch 'branch-A-Gradle'

NaokiShum committed:
Add GUI

NaokiShum committed:
Add GUI

NaokiShum committed:
Adjust code for checkstyle

chownces commented on data/duke.txt:
```diff
@@ -0,0 +1,2 @@
+T | 1 | read book
+D | 1 | return book | Jun 12 2020
```
I feel that the save file should not be committed to the repo! Perhaps, you could try adding `duke/` to your `.gitignore` file and stop Git from tracking this.

https://www.git-tower.com/learn/git/faq/ignore-tracked-files-in-git/

chownces commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,120 @@
+            for (int i = 0; i < tasks.size(); i++) {
+                list += listNum + "." + tasks.get(i) + "\n";
+                listNum++;
+            }
```
Can consider using `StringBuilder` instead of string concatenation. This would allow for performance increase for large strings

chownces commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,120 @@
+        System.out.println(command);
+        switch(command) {
+        case "bye":
+            return ui.showBye();
```
`switch` statements should have a `break;` statement at the end of each `case` block. When a fallthrough is intended, a comment can be added to indicate this to other reviewers/ software engineers - e.g. `// Fallthrough`

Also, you can consider wrapping the entire switch statement in a try-catch block, instead of using it inside every `case` block. I feel that this would improve readability

chownces commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,120 @@
+                if (keyword.strip().isEmpty()) {
+                    throw new DukeException(ui.emptyDescriptionError());
+                }
+                for (int i = 0; i < tasks.size(); i++) {
+                    if (tasks.get(i).getDescription().contains(keyword)) {
+                        findList += tasks.get(i) + "\n";
+                    }
+                }
```
I feel that you can consider abstracting this inside the `TaskList` class itself

chownces commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,67 @@
+     * @param img takes in an img representing the image
+     */
+    public DialogBox(String text, Image img) {
+        try {
+
+            FXMLLoader fxmlLoader = new FXMLLoader(MainWindow.class.getResource("/view/DialogBox.fxml"));
```
Additional newline here which can be removed

chownces commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,35 @@
+        }
+    }
+
+    /**
+     * You should have your own function to generate a response to user input.
+     * Replace this stub with your completed method.
+     */
+    public String getResponse(String input) {
+        Command c = new Command(input);
+        String s = c.execute(tasks, ui, storage);
+        return s;
+    }
```
The JavaDoc comment should be changed!

chownces commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,31 @@
+        return this.at;
+    }
+
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + "(at:" + at + ")";
+    }
```
Can consider using `String.format` to improve readability

chownces commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,72 @@
+        if (Parser.parseCommand(input).equals("bye")) {
+            Timer timer = new Timer();
+            TimerTask task = new TimerTask() {
+                public void run() {
+                    System.exit(0);
+                }
```
In my opinion, `Platform.exit()` might be more suitable in more complex scenarios. But `System.exit()` works here too!

https://stackoverflow.com/questions/46053974/using-platform-exit-and-system-exitint-together

chownces commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,133 @@
+            case "T":
+                String[] todoArr = s.nextLine().trim().split("\\|");
+                Task todo = new Todo(todoArr[2].trim());
+                if (Integer.parseInt(todoArr[1].trim()) == 1) {
+                    todo.markAsDone();
+                }
```
There are many repeated instances of this. Perhaps, you could consider having an alternative constructor that already marks it as done?

chownces commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,139 @@
+ * Ui class provides methods for user interaction
+ */
+public class Ui {
+    private final String line = "_______________________________________________________\n";
```
This can also be made a static constant

chownces commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,139 @@
+     * Shows a welcome text
+     * @return a String representing welcome text
+     */
+    public String showWelcome() {
+        return line + " Hello! I'm Duke\n" + " What can I do for you?\n" + line;
+    }
```
There are many instances of `line + "something" + line`. Perhaps, you could consider having a private method that does that?

chownces commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,11 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class DukeTest {
+    @Test
+    public void dummyTest(){
+        assertEquals(2, 2);
+    }
```
um, this should not be here

chownces commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,13 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class ParserTest {
+    @Test
+    public void Test1(){
+        String test = "BYE";
+        assertEquals("bye", Parser.parseCommand(test));
+    }
```
I feel that the test description can be more descriptive! Similarly for the rest of the tests.

See https://se-education.org/guides/conventions/java/intermediate.html#naming

chownces reviewed (commented):
Overall great work! Feel free to take up the suggestions mentioned and all the best for the rest of the module! :)

NaokiShum committed:
Modify javadocs

NaokiShum committed:
Add Assertions

To document important assumptions that should hold at various points in
the code.

Let's,
* update the execute function to use assertions

NaokiShum committed:
Add javadocs

NaokiShum committed:
Merge pull request #1 from NaokiShum/branch-A-Assertions

Branch a assertions

NaokiShum committed:
Improve code quality

NaokiShum committed:
Merge pull request #2 from NaokiShum/branch-A-CodeQuality

Improve code quality

NaokiShum committed:
Add new task DoAfter

Support the managing of tasks that need to be done after a
specific time/task

Let's,
* update the execute method in command.java to accept a DoAfter task
* add a new DoAfter class
* update the ui to show a response when user uses a DoAfter task

NaokiShum committed:
Enable doAfter task to be saved

NaokiShum committed:
Shorten long method

NaokiShum committed:
Prevent truncation in GUI

NaokiShum committed:
Set theme jekyll-theme-midnight

NaokiShum committed:
Add user guide

NaokiShum committed:
add user guide

NaokiShum committed:
add user guide

NaokiShum committed:
add user guide

NaokiShum committed:
Add user guide

NaokiShum committed:
Add user guide

NaokiShum committed:
Set theme jekyll-theme-hacker

NaokiShum committed:
Add User Guide

NaokiShum committed:
Merge branch 'master' of https://github.com/NaokiShum/ip

NaokiShum committed:
Add User Guide

NaokiShum committed:
Add some helper functions to avoid long methods

NaokiShum committed:
Fix checkstyle

NaokiShum committed:
Move filePath out of the function
